### PR TITLE
Remaining Slasher Beacon Node Changes

### PR DESCRIPTION
**Pull request by rauljordan**
This PR contains remaining changes from feature/slasher to the beacon-chain/ subpackages. The changes including are:

- [x] If --slasher is enabled, forward attestations in blocks to the slasher service in a beacon node during block processing
- [x] Start a slasherDB and slasher service in the beacon node if the --slasher flag is enabled
- [x] RPC implementation of the slasher service with the newly optimized slasher
- [x] If --slasher is enabled, forward attestations and blocks in the sync service from gossipsub to slasher
- [x] Add a --slasher flag
- [x] Remove deprecated slasher protobuf endpoints
- [x] Minor changes in validator to support the new slasher RPC service. **More changes to validator will come in the next PR**

### CODE REVIEW

File: beacon-chain/node/node.go [hunk 1]
```diff
@@ -130,6 +137,12 @@ func New(cliCtx *cli.Context) (*BeaconNode, error) {
 		return nil, err
 	}
 
+	if features.Get().EnableSlasher {
```
Comment (terencechain) [inline beacon-chain/node/node.go:140]: Put the feature flag within `startSlasherDB`?

File: beacon-chain/slasher/rpc.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	attesterSlashings, err := s.checkSlashableAttestations(ctx, []*slashertypes.IndexedAttestationWrapper{indexedAttWrapper})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not check if attestation is slashable: %v", err)
+	}
+	if len(attesterSlashings) == 0 {
+		// If the incoming attestations are not slashable, we mark them as saved in
+		// slasher's DB storage to help us with future detection.
+		if err := s.serviceCfg.Database.SaveAttestationRecordsForValidators(
```
Comment (terencechain) [inline beacon-chain/slasher/rpc.go:56]: Saving to DB seems out of scope base on method name and comment. Can the caller do it?

File: beacon-chain/sync/validate_beacon_attestation.go [hunk 3]
```diff
@@ -81,6 +82,32 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
...
+			indexedAtt, err := attestation.ConvertToIndexed(ctx, att, committee)
+			if err != nil {
+				log.WithError(err).Error("Could not convert to indexed attestation")
+				tracing.AnnotateError(span, err)
+				return
+			}
+			s.cfg.SlasherAttestationsFeed.Send(indexedAtt)
+		}()
+	}
+
```
Comment (terencechain) [inline beacon-chain/sync/validate_beacon_attestation.go:110]: Move this below L140 so you can get `preState` for free

File: config/features/flags.go [hunk 4]
```diff
@@ -93,6 +93,10 @@ var (
 		Name:  "disable-next-slot-state-cache",
 		Usage: "Disable next slot cache which improves attesting and proposing efficiency by caching the next slot state at the end of the current slot",
 	}
+	enableSlasherFlag = &cli.BoolFlag{
+		Name:  "slasher",
+		Usage: "Enables a slasher in the beacon node for detecting slashable offenses on eth2",
```
Comment (terencechain) [inline config/features/flags.go:98]:
```suggestion
		Usage: "Enables a slasher in the beacon node for detecting slashable offenses on beacon chain",
```

File: beacon-chain/sync/validate_beacon_attestation.go (hunk 3 above)
Comment (rauljordan) [reply beacon-chain/sync/validate_beacon_attestation.go:110]: Discussed this with Sproul. Unfortunately, that happens too late in the validation pipeline. We need to send the attestation soon after the nil and timing checks, but before we do other validation in gossip for it.

File: beacon-chain/slasher/rpc.go (hunk 2 above)
Comment (rauljordan) [reply beacon-chain/slasher/rpc.go:56]: We would need to have two RPC endpoints. One for checking an att is slashable and another for committing. I agree that seems like a more sensible option, but for the sake of time, perhaps we can proceed with this and improve once we have an rc?

Comment (terencechain) [reply beacon-chain/slasher/rpc.go:56]: Let's rename the function to reflect it is saving to db?

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -106,6 +107,32 @@ func (s *Service) onBlock(ctx context.Context, signed block.SignedBeaconBlock, b
...
+			for _, att := range signed.Block().Body().Attestations() {
+				committee, err := helpers.BeaconCommitteeFromState(ctx, preState, att.Data.Slot, att.Data.CommitteeIndex)
+				if err != nil {
+					log.WithError(err).Error("Could not get attestation committee")
+					tracing.AnnotateError(span, err)
+					return
+				}
+				// Using a different context to prevent timeouts as this operation can be expensive
+				// and we want to avoid affecting the critical code path.
+				indexedAtt, err := attestation.ConvertToIndexed(context.TODO(), att, committee)
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block.go:125]: If this is the case, wouldn't you also want to use the different context on line 117? 

I suggest shadowing `ctx` after starting the anonymous function.
```go
go func() {
  ...
  ctx := context.TODO()
  ...
```

File: beacon-chain/blockchain/testing/BUILD.bazel [hunk 6]
```diff
@@ -2,7 +2,6 @@ load("@prysm//tools/go:def.bzl", "go_library")
 
 go_library(
     name = "go_default_library",
-    testonly = True,
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/testing/BUILD.bazel:5]: Please revert this so that test code does not leak into production code.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Partial review, still looking

File: beacon-chain/slasher/rpc.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/slasher/rpc.go:56]: I think that saving to the db, if we have never seen this attestation before, is reasonable. 

However, the reason we original had this into two methods is that we didn't want the slasher to broadcast our own slashable events.

File: validator/db/kv/prune_attester_protection_test.go [hunk 7]
```diff
@@ -39,7 +39,7 @@ func TestPruneAttestations_NoPruning(t *testing.T) {
 }
 
 func TestPruneAttestations_OK(t *testing.T) {
-	numKeys := uint64(2048)
+	numKeys := uint64(64)
```
Comment (prestonvanloon) [inline validator/db/kv/prune_attester_protection_test.go:42]: Please explain this change

File: beacon-chain/rpc/service.go [hunk 8]
```diff
@@ -34,8 +34,10 @@ import (
 	beaconv1alpha1 "github.com/prysmaticlabs/prysm/beacon-chain/rpc/prysm/v1alpha1/beacon"
 	debugv1alpha1 "github.com/prysmaticlabs/prysm/beacon-chain/rpc/prysm/v1alpha1/debug"
 	nodev1alpha1 "github.com/prysmaticlabs/prysm/beacon-chain/rpc/prysm/v1alpha1/node"
+	slasher2 "github.com/prysmaticlabs/prysm/beacon-chain/rpc/prysm/v1alpha1/slasher"
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go:37]: Consider renaming to slasherpb or something more canonical. slasher2 seems like a pattern that we are trying to avoid.

File: validator/db/kv/prune_attester_protection_test.go (hunk 7 above)
Comment (rauljordan) [reply validator/db/kv/prune_attester_protection_test.go:42]: Same functionality and thing we are testing. However, 2048 takes a long time for test set up. If we want to benchmark things, we have a benchmark function we can use. The unit test is just to check things work as expected

File: beacon-chain/blockchain/process_block.go [hunk 9]
```diff
@@ -106,6 +107,33 @@ func (s *Service) onBlock(ctx context.Context, signed block.SignedBeaconBlock, b
 		return err
 	}
 
+	// If slasher is configured, forward the attestations in the block via
+	// an event feed for processing.
+	if features.Get().EnableSlasher {
+		// Feed the indexed attestation to slasher if enabled. This action
+		// is done in the background to avoid adding more load to this critical code path.
+		go func() {
+			ctx := context.TODO() // Shadow context as given this is a side-effect we don't want to affect the main thread.
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block.go:116]: I don't think this comment is correct. The reason we are doing this is the same reason you gave on lines 124 and 125.

```
// Using a different context to prevent timeouts as this operation can be expensive
// and we want to avoid affecting the critical code path.
```

Comment (prestonvanloon) [reply beacon-chain/blockchain/process_block.go:116]: and then you dont need that other context.TODO(), both functions can use `ctx`

**Review by prestonvanloon — APPROVED** [review]
